From 2e89f711f976473351ffb54856bc5738de8778aa Mon Sep 17 00:00:00 2001
From: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date: Mon, 4 Jun 2018 13:23:17 +0000
Subject: [PATCH] OS-6937 support large page mapping for bhyve IOMMU domains
 Reviewed by: Patrick Mooney <patrick.mooney@joyent.com> Approved by: John
 Levon <john.levon@joyent.com>

---
 usr/src/uts/i86pc/io/vmm/intel/vtd.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/usr/src/uts/i86pc/io/vmm/intel/vtd.c b/usr/src/uts/i86pc/io/vmm/intel/vtd.c
index 2626d9dbed..d8ad20ee46 100644
--- a/usr/src/uts/i86pc/io/vmm/intel/vtd.c
+++ b/usr/src/uts/i86pc/io/vmm/intel/vtd.c
@@ -705,6 +705,7 @@ vtd_create_domain(vm_paddr_t maxaddr)
 	if ((uintptr_t)dom->ptp & PAGE_MASK)
 		panic("vtd_create_domain: ptp (%p) not page aligned", dom->ptp);
 
+#ifdef __FreeBSD__
 #ifdef notyet
 	/*
 	 * XXX superpage mappings for the iommu do not work correctly.
@@ -722,6 +723,18 @@ vtd_create_domain(vm_paddr_t maxaddr)
 	 */
 	dom->spsmask = VTD_CAP_SPS(vtdmap->cap);
 #endif
+#else
+	/*
+	 * On illumos we decidedly do not remove memory mapped to a VM's domain
+	 * from the host_domain, so we don't have to deal with page demotion and
+	 * can just use large pages.
+	 *
+	 * Since VM memory is currently allocated as 4k pages and mapped into
+	 * the VM domain page by page, the use of large pages is essentially
+	 * limited to the host_domain.
+	 */
+	dom->spsmask = VTD_CAP_SPS(vtdmap->cap);
+#endif
 
 	SLIST_INSERT_HEAD(&domhead, dom, next);
 
-- 
2.21.0

